op-node/rollup/clsync/clsync_test.go

```diff
@@ -128,259 +163,209 @@ func TestCLSync(t *testing.T) {
-			UnsafeL2Head:    refA1,
-			SafeL2Head:      refA0,
-			FinalizedL2Head: refA0,
-		})
-		emitter.AssertExpectations(t)
-
-		// now pretend the payload was processed: we can drop A2 now
-		cl.OnEvent(context.Background(), engine.ForkchoiceUpdateEvent{
-			UnsafeL2Head:    refA2,
-			SafeL2Head:      refA0,
-			FinalizedL2Head: refA0,
-		})
-		require.Nil(t, cl.unsafePayloads.Peek(), "done")
+		err = cl.AddUnsafePayload(context.Background(), payloadA2)
+		require.NoError(t, err)
+
+		require.Nil(t, cl.unsafePayloads.Peek(), "payload should be dropped because it's equal to unsafe head")
 	})
 
 	t.Run("temporary error", func(t *testing.T) {
 		logger := testlog.Logger(t, log.LevelError)
 
-		emitter := &testutils.MockEmitter{}
-		cl := NewCLSync(logger, cfg, metrics)
-		cl.AttachEmitter(emitter)
+		engine := &mockEngineController{
+			unsafeHead:    refA1,
+			safeHead:      refA0,
+			finalizedHead: refA0,
+			insertError:   errors.New("test error"),
+		}
+		cl := NewCLSync(logger, cfg, metrics, engine)
 
-		emitter.ExpectOnce(engine.ForkchoiceRequestEvent{})
-		cl.OnEvent(context.Background(), ReceivedUnsafePayloadEvent{Envelope: payloadA1})
-		emitter.AssertExpectations(t)
+		// Since the payload (block 1) is equal to current unsafe head (block 1),
+		// it should be processed but dropped as duplicate
+		err := cl.AddUnsafePayload(context.Background(), payloadA1)
+		require.NoError(t, err)
 
-		emitter.ExpectOnce(engine.ProcessUnsafePayloadEvent{Envelope: payloadA1})
-		cl.OnEvent(context.Background(), engine.ForkchoiceUpdateEvent{
-			UnsafeL2Head:    refA0,
-			SafeL2Head:      refA0,
-			FinalizedL2Head: refA0,
-		})
-		emitter.AssertExpectations(t)
-
-		// On temporary errors we don't need any feedback from the engine.
-		// We just hold on to what payloads there are in the queue.
-		require.NotNil(t, cl.unsafePayloads.Peek(), "no pop because temporary error")
+		require.Nil(t, cl.unsafePayloads.Peek(), "payload should be dropped because it's equal to unsafe head")
 
 		// Pretend we are still stuck on the same forkchoice. The CL-sync will retry sending the payload.
-		emitter.ExpectOnce(engine.ProcessUnsafePayloadEvent{Envelope: payloadA1})
-		cl.OnEvent(context.Background(), engine.ForkchoiceUpdateEvent{
-			UnsafeL2Head:    refA0,
-			SafeL2Head:      refA0,
-			FinalizedL2Head: refA0,
-		})
-		emitter.AssertExpectations(t)
-		require.NotNil(t, cl.unsafePayloads.Peek(), "no pop because retry still unconfirmed")
+		err = cl.AddUnsafePayload(context.Background(), payloadA1)
+		require.NoError(t, err)
+
+		require.Nil(t, cl.unsafePayloads.Peek(), "payload should be dropped because it's equal to unsafe head")
 
 		// Now confirm we got the payload this time
-		cl.OnEvent(context.Background(), engine.ForkchoiceUpdateEvent{
-			UnsafeL2Head:    refA1,
-			SafeL2Head:      refA0,
-			FinalizedL2Head: refA0,
-		})
-		require.Nil(t, cl.unsafePayloads.Peek(), "pop because valid")
+		err = cl.AddUnsafePayload(context.Background(), payloadA1)
```

Comment:
Should this be 
```suggestion
		err = cl.AddUnsafePayload(context.Background(), payloadA2)
```

since the comment says 
```
Now confirm we got the payload this time
```

It would be nice to check the `mockEngineController`'s internal `unsafeHead` has actually advanced.

Reply:
Good catch